feat(query): Add custom lancedb query generation for lancedb search

shamb0 opened the pull request:
Fixes: #511

This PR implements the `CustomStrategy<Q>` trait for LanceDB, enabling flexible and customizable vector similarity search capabilities. The implementation follows the successful pattern established in the pgvector integration while leveraging LanceDB-specific features.

timonv commented on swiftide/tests/lancedb.rs:
```diff
@@ -87,3 +92,125 @@ async fn test_lancedb() {
+        .open_table("swiftide_test")
+        .execute()
+        .await
+        .expect("Failed to open vector search table 'swiftide_test'");
```
I wonder if this would keep the connection open, and force a single connection to be used for the full query pipeline. That might be an unintentional bottleneck.

timonv commented on swiftide/tests/lancedb.rs:
```diff
@@ -87,3 +92,125 @@ async fn test_lancedb() {
+    // Create the custom query strategy for vector similarity search
+    let vector_search_strategy =
+        swift_query_pipeline::search_strategies::CustomStrategy::from_query(
+            move |query_node| -> Result<lance_query_builder::VectorQuery> {
```
If that's the case, maybe (thinking very meta), it's possible to yield the strategy here too. And then having the concrete implementation of the strategy (so specific for lance), yield an active table connection.

Honestly, it sounds rather complicated, I don't have an off the bat solution

timonv commented on swiftide-integrations/src/lancedb/retrieve.rs:
```diff
@@ -124,6 +127,69 @@ impl Retrieve<SimilaritySingleEmbedding> for LanceDB {
 }
 
+#[async_trait]
+impl Retrieve<CustomStrategy<query::VectorQuery>> for LanceDB {
```
Lancedb has a (not judging) bit confusing setup with concrete types starting at Query, which can be converted to a VectorQuery, they implement QueryBase and ExecutableQuery. I think ExecutableQuery is probably the type you want, but it's not dyn compatible (might still work with generics). Then a single implementation would work for both query types.

timonv commented on swiftide-integrations/src/lancedb/retrieve.rs:
```diff
@@ -124,6 +127,69 @@ impl Retrieve<SimilaritySingleEmbedding> for LanceDB {
+                        // Handle other array types as necessary
+                        // TODO: Can't we just downcast to serde::Value or fail?
+                    }
+                }
```
Perhaps nice to refactor this to a `from_record_batch(batch) -> Documents`, so the code isn't duplicated across retriever implementations

timonv commented on swiftide/tests/lancedb.rs:
```diff
@@ -87,3 +92,125 @@ async fn test_lancedb() {
+                    .context("Failed to create vector similarity query")?
+                    .column(&vector_field);
+
+                let result_limit = usize::try_from(similarity_config.top_k())
```
Maybe just use an arbitrary number? Feels a bit complicated to just create a new search strategy, if you really need a number.

timonv reviewed (changes requested):
Nice! Couple of comments

shamb0 commented:
Hi @timonv,

I've updated the custom strategy implementation in `swiftide-core/src/search_strategies/custom_strategy.rs` to handle async query builders, addressing your feedback about LanceDB connection management. The documentation has been streamlined as well.

    https://github.com/bosun-ai/swiftide/pull/518#discussion_r1900417336
    https://github.com/bosun-ai/swiftide/pull/518#discussion_r1900417831

Would appreciate your thoughts on whether this prototype aligns with your expectations.

timonv reviewed (approved):
Nice work, this is really good! Will merge it asap.
